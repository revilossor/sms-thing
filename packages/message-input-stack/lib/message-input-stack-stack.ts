/* eslint-disable no-useless-escape */
import { Stack, StackProps, Aws } from 'aws-cdk-lib'
import { Construct } from 'constructs'
import { Topic } from 'aws-cdk-lib/aws-sns'
import { Role, ServicePrincipal, PolicyDocument, PolicyStatement } from 'aws-cdk-lib/aws-iam'
import { RestApi, AwsIntegration, PassthroughBehavior } from 'aws-cdk-lib/aws-apigateway'

export class MessageInputStack extends Stack {
  private static readonly APIG_SERVICE_PRINCIPAL = 'apigateway.amazonaws.com'

  constructor (scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)

    // TODO make FIFO
    const topic = new Topic(this, 'SendMessageTopic')

    const gatewayExecutionRole: any = new Role(this, 'GatewayExecutionRole', {
      assumedBy: new ServicePrincipal(MessageInputStack.APIG_SERVICE_PRINCIPAL),
      inlinePolicies: {
        PublishMessagePolicy: new PolicyDocument({
          statements: [new PolicyStatement({
            actions: ['sns:Publish'],
            resources: [topic.topicArn]
          })]
        })
      }
    })

    const api = new RestApi(this, 'MessagingApi')

    // TODO validation?
    api.root.addMethod('POST',
      new AwsIntegration({
        service: 'sns',
        integrationHttpMethod: 'POST',
        path: `${Aws.ACCOUNT_ID}/${topic.topicName}`,
        options: {
          credentialsRole: gatewayExecutionRole,
          passthroughBehavior: PassthroughBehavior.NEVER,
          requestParameters: {
            'integration.request.header.Content-Type': '\'application/json\''
          },
          requestTemplates: {
            'application/json': `Action=Publish&TopicArn=$util.urlEncode('${topic.topicArn}')&Message=$util.urlEncode($input.body)`
          },
          integrationResponses: [
            {
              statusCode: '200',
              responseTemplates: {
                'application/json': '{"status": "message received"}'
              }
            },
            {
              statusCode: '400',
              selectionPattern: '^\[Error\].*',
              responseTemplates: {
                'application/json': '{\"state\":\"error\",\"message\":\"$util.escapeJavaScript($input.path(\'$.errorMessage\'))\"}'
              }
            }
          ]
        }
      }),
      {
        methodResponses: [
          { statusCode: '200' },
          { statusCode: '400' }
        ]
      }
    )
  }
}
