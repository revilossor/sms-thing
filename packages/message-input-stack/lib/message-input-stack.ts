/* eslint-disable no-useless-escape */
import { Stack, StackProps, Aws, CfnOutput } from 'aws-cdk-lib'
import { Construct } from 'constructs'
import { Topic } from 'aws-cdk-lib/aws-sns'
import { Role, ServicePrincipal, PolicyDocument, PolicyStatement } from 'aws-cdk-lib/aws-iam'
import { RestApi, AwsIntegration, PassthroughBehavior, RequestValidator, Model, JsonSchemaType } from 'aws-cdk-lib/aws-apigateway'

export class MessageInputStack extends Stack {
  private static readonly APIG_SERVICE_PRINCIPAL = 'apigateway.amazonaws.com'

  constructor (scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)

    // TODO maybe api key / rate limit?

    const topic = new Topic(this, 'SendMessageTopic', {
      fifo: true,
      contentBasedDeduplication: true,
      topicName: 'send-message-topic'
    })

    const gatewayExecutionRole: any = new Role(this, 'GatewayExecutionRole', {
      assumedBy: new ServicePrincipal(MessageInputStack.APIG_SERVICE_PRINCIPAL),
      inlinePolicies: {
        PublishMessagePolicy: new PolicyDocument({
          statements: [new PolicyStatement({
            actions: ['sns:Publish'],
            resources: [topic.topicArn]
          })]
        })
      }
    })

    const api = new RestApi(this, 'MessagingApi')

    const sendMessageModel: Model = api.addModel('SendMessageModel', {
      schema: {
        type: JsonSchemaType.OBJECT,
        properties: {
          message: {
            type: JsonSchemaType.STRING // TODO min/max message length?
          },
          phoneNumber: {
            type: JsonSchemaType.STRING,
            minLength: 11,
            maxLength: 11
          }
        },
        required: ['message', 'phoneNumber'],
        additionalProperties: false
      }
    })

    api.root.addMethod('POST',
      new AwsIntegration({
        service: 'sns',
        integrationHttpMethod: 'POST',
        path: `${Aws.ACCOUNT_ID}/${topic.topicName}`,
        options: {
          credentialsRole: gatewayExecutionRole,
          passthroughBehavior: PassthroughBehavior.NEVER,
          requestParameters: {
            'integration.request.header.Content-Type': '\'application/json\''
          },
          requestTemplates: {
            'application/json': `Action=Publish&TopicArn=$util.urlEncode('${topic.topicArn}')&Message=$util.urlEncode($input.body)`
          },
          integrationResponses: [
            {
              statusCode: '200',
              responseTemplates: {
                'application/json': '{"status": "message received"}'
              }
            },
            {
              statusCode: '400',
              selectionPattern: '^\[Error\].*',
              responseTemplates: {
                'application/json': '{\"state\":\"error\",\"message\":\"$util.escapeJavaScript($input.path(\'$.errorMessage\'))\"}'
              }
            }
          ]
        }
      }),
      {
        methodResponses: [
          { statusCode: '200' },
          { statusCode: '400' }
        ],
        requestValidator: new RequestValidator(
          this,
          'RequestValidator',
          {
            restApi: api,
            requestValidatorName: 'sendMessageBodyValidator',
            validateRequestBody: true,
            validateRequestParameters: false
          }
        ),
        requestModels: {
          'application/json': sendMessageModel
        }
      }
    )

    // eslint-disable-next-line no-new
    new CfnOutput(this, 'SendMessageTopicARN', { value: topic.topicArn })
  }
}
