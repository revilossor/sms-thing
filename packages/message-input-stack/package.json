{
  "name": "message-input-stack",
  "version": "0.1.0",
  "repository": {
    "type": "git",
    "url": "git@github.com:revilossor/sms-thing.git"
  },
  "scripts": {
    "deploy": "cdk deploy -- --require-approval never",
    "test": "jest"
  },
  "dependencies": {
    "aws-cdk-lib": "2.16.0",
    "constructs": "^10.0.0",
    "source-map-support": "^0.5.16"
  },
  "devDependencies": {
    "@types/jest": "^26.0.10",
    "@types/node": "10.17.27",
    "aws-cdk": "2.16.0",
    "jest": "^26.4.2",
    "ts-jest": "^26.2.0",
    "ts-node": "^9.0.0",
    "typescript": "~3.9.7"
  }
}
