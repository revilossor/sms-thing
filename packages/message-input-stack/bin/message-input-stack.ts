#!/usr/bin/env node
/* eslint-disable no-new */
import 'source-map-support/register'
import * as cdk from 'aws-cdk-lib'
import { MessageInputStack } from '../lib/message-input-stack-stack'

const app = new cdk.App()

new MessageInputStack(app, 'MessageInputStack', {})
