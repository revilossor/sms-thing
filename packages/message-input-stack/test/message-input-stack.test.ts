import * as cdk from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import { MessageInputStack } from '../lib/message-input-stack'

it('Stack matches snapshot', () => {
  const app = new cdk.App()
  const stack = new MessageInputStack(app, 'TestMessageInputStack')
  const template = Template.fromStack(stack)
  expect(template.toJSON()).toMatchSnapshot()
})
