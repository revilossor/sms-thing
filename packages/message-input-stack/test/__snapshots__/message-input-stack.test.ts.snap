// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message input stack matches the snapshot 1`] = `
Object {
  "Outputs": Object {
    "MessagingApiEndpoint78487A17": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "MessagingApi6D1F8F71",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "MessagingApiDeploymentStageprod64BAE454",
            },
            "/",
          ],
        ],
      },
    },
    "SendMessageTopicARN": Object {
      "Value": Object {
        "Ref": "SendMessageTopic124D0AB8",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "GatewayExecutionRole16B5E8DF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "GatewayExecutionRoleDefaultPolicyD642E70C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "SendMessageTopic124D0AB8",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GatewayExecutionRoleDefaultPolicyD642E70C",
        "Roles": Array [
          Object {
            "Ref": "GatewayExecutionRole16B5E8DF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MessagingApi6D1F8F71": Object {
      "Properties": Object {
        "Name": "MessagingApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "MessagingApiAccount295EFC8E": Object {
      "DependsOn": Array [
        "MessagingApi6D1F8F71",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "MessagingApiCloudWatchRole9E703F5A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "MessagingApiCloudWatchRole9E703F5A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MessagingApiDeploymentDDE83BBC40e8368e4025f3bae252001a772de97f": Object {
      "DependsOn": Array [
        "MessagingApiPOSTDA209080",
        "MessagingApiSendMessageModelB218FB47",
        "RequestValidatorB6FDBF18",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "MessagingApi6D1F8F71",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "MessagingApiDeploymentStageprod64BAE454": Object {
      "DependsOn": Array [
        "MessagingApiAccount295EFC8E",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "MessagingApiDeploymentDDE83BBC40e8368e4025f3bae252001a772de97f",
        },
        "RestApiId": Object {
          "Ref": "MessagingApi6D1F8F71",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "MessagingApiPOSTDA209080": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "GatewayExecutionRole16B5E8DF",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "ResponseTemplates": Object {
                "application/json": "{\\"status\\": \\"message received\\"}",
              },
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "{\\"state\\":\\"error\\",\\"message\\":\\"$util.escapeJavaScript($input.path('$.errorMessage'))\\"}",
              },
              "SelectionPattern": "^[Error].*",
              "StatusCode": "400",
            },
          ],
          "PassthroughBehavior": "NEVER",
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/x-www-form-urlencoded'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "Action=Publish&TopicArn=$util.urlEncode('",
                  Object {
                    "Ref": "SendMessageTopic124D0AB8",
                  },
                  "')&Message=$util.urlEncode($input.body)&MessageGroupId=api",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":sns:path//",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "StatusCode": "200",
          },
          Object {
            "StatusCode": "400",
          },
        ],
        "RequestModels": Object {
          "application/json": Object {
            "Ref": "MessagingApiSendMessageModelB218FB47",
          },
        },
        "RequestValidatorId": Object {
          "Ref": "RequestValidatorB6FDBF18",
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "MessagingApi6D1F8F71",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "MessagingApi6D1F8F71",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "MessagingApiSendMessageModelB218FB47": Object {
      "Properties": Object {
        "ContentType": "application/json",
        "RestApiId": Object {
          "Ref": "MessagingApi6D1F8F71",
        },
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "additionalProperties": false,
          "properties": Object {
            "message": Object {
              "maxLength": 140,
              "type": "string",
            },
            "phoneNumber": Object {
              "maxLength": 11,
              "minLength": 11,
              "type": "string",
            },
          },
          "required": Array [
            "message",
            "phoneNumber",
          ],
          "type": "object",
        },
      },
      "Type": "AWS::ApiGateway::Model",
    },
    "RequestValidatorB6FDBF18": Object {
      "Properties": Object {
        "Name": "sendMessageBodyValidator",
        "RestApiId": Object {
          "Ref": "MessagingApi6D1F8F71",
        },
        "ValidateRequestBody": true,
        "ValidateRequestParameters": false,
      },
      "Type": "AWS::ApiGateway::RequestValidator",
    },
    "SendMessageTopic124D0AB8": Object {
      "Properties": Object {
        "ContentBasedDeduplication": true,
        "FifoTopic": true,
        "TopicName": "send-message-topic.fifo",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
